app/src/main/java/org/merakilearn/ui/onboarding/WelcomeFragment.kt

```diff
@@ -74,6 +80,63 @@ class WelcomeFragment : BaseFragment() {
         })
     }
 
+    private fun initGoogleSignInOption() {
+        val gso =
+            GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
+                .requestEmail()
+                .requestIdToken(getString(R.string.server_client_id))
+                .build()
+        mGoogleSignInClient = GoogleSignIn.getClient(requireActivity(), gso)
+    }
+
+    private fun signInWithGoogle() {
+        val signInIntent = mGoogleSignInClient!!.signInIntent
+        startActivityForResult(signInIntent, RC_SIGN_IN)
+    }
+
+
+    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
+        super.onActivityResult(requestCode, resultCode, data)
+        if (requestCode == RC_SIGN_IN) {
+            val task =
+                GoogleSignIn.getSignedInAccountFromIntent(data)
+            handleSignInResult(task)
+        }
+    }
+
+    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
+        try {
+            val account = completedTask.getResult(ApiException::class.java)
+            logOutSelectedAccount(account)
+        } catch (e: ApiException) {
+            Log.e(LoginFragment.TAG, "signInResult:failed code=", e)
+            toast(getString(R.string.unable_to_sign))
+        }
+    }
+
+    private fun logOutSelectedAccount(account: GoogleSignInAccount) {
+        mGoogleSignInClient?.signOut()?.addOnSuccessListener {
+            viewModel.idToken.value = account.idToken
+            observeFetchingBackendToken()
+        }
+    }
+
+    private fun observeFetchingBackendToken() {
+        toggleProgressBarVisibility(View.VISIBLE)
+        viewModel.loginResult.observe(this, Observer {
```

Comment:
Observers should not be registered multiple times. They should be register once. If we register on some click event, they will be registered everytime we click on it

Reply:
Will remove from click event